www-reputation-credentials-protocol.md

```diff
@@ -0,0 +1,265 @@
+
+## Checking HTLC forward request
+
+Accepting and forwarding a payment as described in BOLT 4 is modified. A routing hop requiring reputation credentials
+should enforce additional checks.
+
+```
+				     
+					  2.	
+					      routing_policy(credentials)? 
+	 _________   1.	                              _________		                               __________
+	|	  |	  	                     |	       | 3a.		                      |	         |
+	|	  |    HTLC + onions(credentials)    |	       |    yes           HTLC   	      |	         |
+	|  Alice  |--------------------------------->|   Bob   |------------------------------------->|  Caroll  |
+	|	  |	                     	     |	       |		   	              |	         |
+	|_________|	                     3b.     |_________|		                      |__________|
+		^				no    /
+		 \	     fail_htlc		     /
+		  \_________________________________/
+			
+```
+
+In this diagram:
+1. Alice sends a HTLC+credentials to Bob.
+2. Bob extracts the credentials, verifies the signature and validate the credentials against his own `routing_policy`.
+3a. The routing checks are successful, Bob forwards the HTLC along the Bob-Caroll link.
+4a. Bob stores the backward credentials into the local state until the HTLC settlement.
+3b. The routing checks are unsuccessful, Bob rejects the HTLC.
+
+The credentials are wrapped inside the HTLC onion `credentials_payload`. There are two types of credentials: forward/backward.
+The forward credentials are covering the ongoing HTLC forward, they should have been counter-signed by the target routing hop
+during the dissemination phase or a previous HTLC forward/settlement pahse with the same target routing hop. The backward
+credentials are also wrapped into the HTLC onion, however they're not signed and should be blinded. If the settlement is successful,
+those backward credentials should be the "forward" credentials for Alice's future HTLC sends.
+
+## Signing or slashing credentials during HTLC Settlement
+
+The current HTLC settlement does not encompass any routing checks processing by routing hops, whatever the outcome `update_fulfill_htlc`/
+`update_fail_htlc`.
+
+```
+						    2. HTLC success ?
+
+							2a. yes, sign backward credentials
+
+							2b. no, slash forward credentials
+
+	 _________   3a. HTLC + onions(signed_credentials)   _________						__________
+	|	  |<----------------------------------------|	      |	       1.			       |	  |
+	|	  |				            |	      |		  HTLC settlement	       |	  |
+	|  Alice  |				            |   Bob   |<---------------------------------------|  Caroll  |
+	|	  |         3b. HTLC 		            |	      |				               |	  |
+	|_________|<----------------------------------------|_________|					       |__________|
+
+```
+
+In this diagram:
+1. Caroll settles the HTLC back to Bob with either a success or a failure.
+2a. If it's a success, Bob sign the blinded backward credentials stored during the previous phase.
```

Comment:
What happens if upstream nodes along the path corrupt the signed credentials somehow? They are part of the `update_fulfill_htlc` message, right?

Reply:
Yes I think the description should be fixed, the signed credentials should be wrapped into an onion. The onion can be attached to the `update_fulfill_htlc` or they could send back with a reply_path as the onion message proposal introduces. This reply_path could have been exchanged during the dissemination phase. I still have to catchup with the latest state of onion messages to see what works the best.

Reply:
Attached to `update_fulfill_htlc` as an onion, I'd think that nodes can still delete it? And stand-alone onion messages may not arrive either. If there is a way to somehow atomically deliver upstream together with the preimage, that seems to be better. Or use something like https://github.com/lightning/bolts/pull/1044 so that the sender can penalize the node that deleted the in-flight credentials.

Reply:
That's a good point, stand-alone onion messages are not reliable. I think effectively you would like the credentials to be part of the hmac chain as proposed by #1044  and the sender should be able to detect tampering. Especially, if the credentials have a satoshi-cost, you open the way to damage by intermediary nodes.

Reply:
Nice to have another reason to work on #1044 😉